### PR TITLE
IRGen: Weakly link associated conformance descriptors when either the protocol or the associated requirement is weak

### PR DESCRIPTION
**Pull request by tshortli**
Previously, only the associated requirement was considered when deciding whether to weakly link an associated conformance descriptor. This lead to unexpected strong linkage for some symbols, interfering with back deployment when integrating with some frameworks.

Resolves rdar://96974850

### CODE REVIEW

**Comment by tshortli** [comment]
@swift-ci please smoke test

**Review by slavapestov — APPROVED** [review]
Nice catch.

**Review by nkcsgexi — APPROVED** [review]
Thank you for fixing it!

**Comment by tshortli** [comment]
macOS test failure is unrelated:
```
Failed Tests (1):
  lldb-api :: functionalities/asan/swift/TestAsanSwift.py
```

Linux test failures also seem unrelated, though they are a bit hard to parse.

**Comment by tshortli** [comment]
@swift-ci please smoke test

**Comment by tshortli** [comment]
@swift-ci please smoke test macOS

**Comment by tshortli** [comment]
@swift-ci please smoke test macOS
